frontend/templates/product-list/index.tsx

```diff
@@ -0,0 +1,279 @@
+import {
+   AppProviders,
+   AppProvidersProps,
+   Layout,
+   LayoutProps,
+   WithLayoutProps,
+   WithProvidersProps,
+} from '@components/common';
+import { ProductListView } from '@components/product-list';
+import { ALGOLIA_PRODUCT_INDEX_NAME } from '@config/env';
+import {
+   decodeDeviceItemType,
+   decodeDeviceTitle,
+   encodeDeviceTitle,
+} from '@helpers/product-list-helpers';
+import { assertNever, invariant, logAsync } from '@ifixit/helpers';
+import { urlFromContext } from '@ifixit/helpers/nextjs';
+import { getGlobalSettings } from '@models/global-settings';
+import {
+   findProductList,
+   ProductList,
+   ProductListType,
+} from '@models/product-list';
+import { getStoreByCode, getStoreList } from '@models/store';
+import { GetServerSideProps, GetServerSidePropsContext } from 'next';
+import { ParsedUrlQuery } from 'querystring';
+import { getServerState } from 'react-instantsearch-hooks-server';
+
+export type ProductListTemplateProps = WithProvidersProps<
+   WithLayoutProps<{
+      productList: ProductList;
+      indexName: string;
+   }>
+>;
+
+export const ProductListTemplate: NextPageWithLayout<ProductListTemplateProps> =
+   ({ productList, indexName }) => {
+      return (
+         <ProductListView productList={productList} indexName={indexName} />
+      );
+   };
+
+ProductListTemplate.getLayout = function getLayout(page, pageProps) {
+   return <Layout {...pageProps.layoutProps}>{page}</Layout>;
+};
+
+export default ProductListTemplate;
+
+type GetProductListServerSidePropsOptions = {
+   productListType: ProductListType;
+};
+
+export const getProductListServerSideProps = ({
+   productListType,
+}: GetProductListServerSidePropsOptions): GetServerSideProps<ProductListTemplateProps> => {
+   return async (context) => {
+      context.res.setHeader(
+         'Cache-Control',
+         'public, s-maxage=10, stale-while-revalidate=600'
+      );
+
+      const indexName = ALGOLIA_PRODUCT_INDEX_NAME;
+      let layoutProps: LayoutProps;
+      let productList: ProductList | null;
+      let shouldRedirectToCanonical = false;
+      let canonicalPath: string | null = null;
+
+      switch (productListType) {
+         case ProductListType.AllParts: {
```

Comment:
I appreciate the de-duplication! It seems like moving the differentiating code into a 120-line switch statement is a different kind of anti-pattern though.

Reply:
@danielbeardsley If you have an improvement in mind, feel free to suggest it. I appreciate short switches too, but this is a big improvement over what we had before.

Reply:
Yeah, I would probably have left the bits that differ across pages in their respective `pages/*` files. That said, maybe this would have resulted in more overhead, I don't know. I commented cause it seems like this swung the pendulum too far the other direction.

Reply:
I thought about that too, putting reusable pieces in a separate folder works too (e.g. `lib`). Maybe it's also possible to transform `getProductListServerSideProps` into a higher order function, so you can move the customization back into the `/pages` folder, but I didn't want to spend too much time thinking about that. I'm sure that as we move forward and accumulate use cases we'll have a better idea on how to improve this.